src/google/protobuf/compiler/csharp/csharp_enum.cc

```diff
@@ -64,11 +65,24 @@ void EnumGenerator::Generate(io::Printer* printer) {
                  "access_level", class_access_level(),
                  "name", descriptor_->name());
   printer->Indent();
+  std::set<string> used_names;
   for (int i = 0; i < descriptor_->value_count(); i++) {
       WriteEnumValueDocComment(printer, descriptor_->value(i));
-      printer->Print("$name$ = $number$,\n",
-                   "name", descriptor_->value(i)->name(),
-                   "number", SimpleItoa(descriptor_->value(i)->number()));
+      string original_name = descriptor_->value(i)->name();
+      string name = options()->legacy_enum_values
+          ? descriptor_->value(i)->name()
+          : GetEnumValueName(descriptor_->name(), descriptor_->value(i)->name());
+      // Make sure we don't get any duplicate names due to prefix removal.
+      while (!used_names.insert(name).second) {
+        // It's possible we'll end up giving this warning multiple times, but that's better than not at all.
+        GOOGLE_LOG(WARNING) << "Duplicate enum value " << name << " (originally " << original_name
+          << ") in " << descriptor_->name() << "; adding underscore to distinguish";
+        name += "_";
```

Comment:
Well, in an ideal world we'd never get into this state. With a universal check that after:

- Prefix removal
- Underscore removal
- Upper-casing

... the enum value names are all unique, we wouldn't need this check. For the moment, I'm not sure of a better solution though. The options I can see are:

- Some sort of munging as per this PR
- Generating invalid code
- Failing with an error when the duplicate is noted

I'd be happy enough to fail, but would welcome advice on the right way to do so from deep within a plugin. Other suggestions very welcome too, certainly.